### PR TITLE
[Bug] FP32 RMSNorm layout fix

### PR DESCRIPTION
**Pull request by bohnstingl**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes a layout-propagation bug that makes a staggered dtype conversion **silently
produce wrong results** for some shapes, and hard-fail at codegen for others.

`_single_arg_op_layout` handles a dtype conversion that creates or consumes a
staggered element arrangement (`DL16_TO_FP32` / `FP32_TO_DL16`) by inheriting the
*input buffer's* `device_size` / `stride_map` and rescaling only the stick depth
(`rescale_stl_for_dtype`). That is correct — and necessary — when the conversion
traverses its whole input: the staggered ordering is physical, existing only on
the device layout, so rebuilding from the logical host size would lose the stick
choice a downstream reduction depends on.

It is **not** correct when the conversion reads a *slice* of a wider buffer. Then
the inherited row span describes the input buffer while the elements land in an
output buffer with a different row stride. `compute_coordinates` folds that
mismatch into the outer (non-stick) coordinate as `Mod(a*var, b)`, where
`a / b = W_out / W_in` in lowest terms. Two outcomes, both bad:

| `a` | Outcome |
| --- | --- |
| `a != 1` | Outside the coordinate-normalization grammar. Hard `Unsupported` from `normalize_var_expr`, raised late, during `SpyreKernel.codegen_kernel`. |
| `a == 1` | *Representable* — so it compiles, and then addresses the wrong sticks. No diagnostic, no warning, wrong numbers. |

This is reached by vLLM's native FP32 RMSNorm on Gemma. Gemma-3/4 `q_norm` and
`k_norm` apply a per-head RMSNorm to a slice of the fused QKV projection output,
so the fp16→fp32 upcast at the head of the norm reads `W_qkv`-wide rows and
writes a narrower dense buffer:

| Case | `W_out / W_in` | Coordinate | Symptom |
| --- | --- | --- | --- |
| gemma-3-1b `q_norm` | 1024/1536 = 2/3 | `Mod(2*c0, 3)` | compile error |
| gemma-3-1b `k_norm` | 256/1536 = 1/6 | `Mod(c0, 6)` | **silent, maxabsdiff 1799** |
| gemma-4-31B `q_norm` | 16384/20480 = 4/5 | `Mod(4*c0, 5)` | compile error |
| gemma-4-31B `k_norm` | 2048/20480 = 1/10 | `Mod(c0, 10)` | **silent** |
| gemma-4-26B-A4B `q_norm` | 8192/10240 = 4/5 | `Mod(4*c0, 5)` | compile error |

Concretely for gemma-3-1b `q_norm`: input `buf10` `[8, 1536]` fp16 has
`device_size=[24, 8, 64]`, `stride_map=[64, 1536, 1]`. Output `buf13` `[8, 4, 256]`
fp32 has host row stride 1024. The inherited-and-rescaled layout is
`device_size=[48, 8, 32]`, `stride_map=[32, 1536, 1]` — claiming 1536-element rows
for a 1024-wide buffer. With `c0` step 1024, extent 8192 and `next_stride[0] = 1536`,
`compute_coordinates` emits
`[8*c1 + 16*Mod(2*c0, 3) + floor(c2/32), floor(2*c0/3), Mod(c2, 32)]`, and
`normalize_var_expr` rejects the `Mod(2*c0, 3)` term.

**The fix**, in two parts:

1. `propagate_layouts.py` — new `_convert_reads_whole_input()` gates the inherit
   path on the conversion traversing its input exactly as it writes its output
   (same host size, same `dep.index`, same `host_coordinates`). This mirrors the
   identical-access test the general non-EA convert path already applies, minus
   its element-width condition — rescaling the stick depth is precisely what this
   path exists for. A sliced read now falls through to the pre-existing dense
   reconstruction from the output host size, which still stamps the staggered EA.

2. `pass_utils.py` — new `coordinates_normalizable()` applies
   `normalize_coordinates` to a candidate layout as a predicate. `device_coordinates`
   validates **only** the stick expression (`_check_stick_expr_supported(coords[-1], …)`);
   non-stick dims are first checked much later, in codegen. This helper closes
   that gap so a doomed candidate is dropped while alternatives are still
   reachable. It is applied as a secondary, last-resort filter in the staggered
   branch — the structural guard in (1) is what actually fixes the defect, and (2)
   would not have caught the silent `a == 1` cases on its own.

Why the dense fallback is physically sound: the staggered EA records that one
fp16 stick (64 lanes) converts into two *adjacent* fp32 sticks. For gemma-3-1b
`q_norm` (`W_q = 1024 = 16 × 64`), input stick `s = 24t + j` with `j ∈ [0, 16)`
maps to output sticks `(32t + 2j, 32t + 2j + 1)` — adjacent, and on the same
device dim (`dim0`, host stride 32) that carried the pairing in the inherited
layout. So the dense reconstruction plus the staggered EA preserves everything
the inherited geometry was there to preserve, without importing the wrong row
span. Verified on paper and numerically.

#### Which issue(s) this PR is related to:

Related to <https://github.com/torch-spyre/torch-spyre/pull/4238>, which addressed
the adjacent restickify-reachability problem. That PR added `spyre::to_dtype_d2d`,
an FP16-only restickify guard, and extra restickify-reachable candidates — but
every candidate it adds is still `rescale_stl_for_dtype(target_stl, …)`, derived
from the *input buffer's* geometry. It widens the choice of **stick axis** and
never questions the **row span**, so the sliced case is untouched by it. This PR
is complementary, not a replacement.

Unblocks the Gemma entries in spyre-inference's
`tests/e2e/test_compile.py::test_basic_llm_inference` and
`tests/e2e/test_distributed_tp2.py`.

#### Special notes for your reviewer:

**Please read this as a silent-wrong-results fix, not only a compile fix.** The
`a == 1` rows in the table above compile cleanly on `main` today and return
garbage. Any Gemma accuracy number measured with vLLM's native FP32 RMSNorm
before this patch is invalid. The `k` cases in the new test are the guard for
exactly that, which is why they assert values — a compile-only check passes even
when every access is misplaced.

**Test results.**

New regression test, `tests/inductor/test_sliced_staggered_convert.py`
(gemma-3-1b and gemma-4-global geometries × `q`/`k`, values asserted against
eager):

```
$ pytest tests/inductor/test_sliced_staggered_convert.py -q
4 passed in 30.59s
```

Standalone reproducer over 8 combinations (gemma3-1b, gemma4-global × q/k/qk):

- **Before:** `gemma3-1b[q]`, `gemma3-1b[qk]` compile-fail `Mod(2*c0, 3)`;
  `gemma4-global[*]` compile-fail `Mod(4*c0, 5)`; `gemma3-1b[k]` compiles and
  returns maxabsdiff **1799** vs eager.
- **After:** all 8 compile and match eager, maxabsdiff 0.0078–0.0195 (fp16 round-trip).

Targeted suites for the code this touches:

```
$ pytest tests/inductor/test_ea_bidirectional_conversion.py tests/inductor/test_coordinates.py
97 passed, 9 subtests passed

$ pytest tests/inductor/test_restickify.py \
         tests/inductor/test_inductor_normalization_scalars.py \
         tests/inductor/test_inductor_dtype_scalars.py
2 failed, 438 passed, 30 skipped, 102 xfailed, 3 xpassed
```

Both failures — `TestNegativeScalarOperations::test_requires_grad_scalar_operation[compiled]`
and `::test_chained_scalar_multiply_twice[compiled]` — are **pre-existing**,
confirmed by stashing this patch and re-running that class on unmodified
torch-spyre: identical two failures.

End-to-end, in spyre-inference against this patched torch-spyre:

```
tests/e2e/test_compile.py::test_basic_llm_inference[model_ref_output1]   # gemma-3-1b
1 passed in 105.30s

tests/models/language/generation/test_gemma.py::test_dummy_loader[google/gemma-3-1b-it]
1 passed in 94.59s
```

**Verification gap, stated plainly.** The `4/5` geometry is validated standalone
and at unit level, but **not** end-to-end on the real `google/gemma-4-31B` and
`google/gemma-4-26B-A4B` checkpoints. `test_basic_llm_inference[model_ref_output2]`,
`[model_ref_output3]` and `test_tp2_compiled_llm_generate_matches_tp1[google/gemma-4-26B-A4B]`
still need a run on those weights to confirm nothing else is in the way. A
dummy-weight 26B build was attempted and was still in weight init after 15
minutes; it was stopped in favour of the standalone validation, which isolates
this defect precisely but says nothing about the rest of those models.

**Out of scope: the mixed-EA join.** A separate defect remains open, and this PR
does not touch it. Granite's `RMSNorm.forward_native` downcasts fp32→fp16
*before* the weight multiply, so that multiply joins a `FP32_TO_DL16` activation
with a `STANDARD` fp16 weight and neither operand broadcasts in the stick
dimension — case 3.3 of `_multi_arg_pointwise_layouts`
(`propagate_layouts.py:1457`), which raises:

```
Unsupported: Spyre backend does not support: Multi-arg pointwise with mixed EA:
STANDARD input arg1_1 must broadcast (device stick dimension size 1) to be
compatible with a staggered EA, and no staggered operand is broadcastable
either. Its stick maps to host dim 0 of size 4096
```

That needs the device-resident EA-conversion op the code comment there already
anticipates ("future: insert an explicit EA conversion at extra cost"). Gemma
never reaches it: `GemmaRMSNorm` multiplies by `1.0 + weight.float()` in fp32, so
both operands come out `DL16_TO_FP32`, which is why Gemma is fixed by this PR
while Granite's native path still needs the custom `SpyreRMSNorm` op. The
`Unsupported` above is raised during `LLM(...)` construction, so it is loud rather
than silent — no correctness risk, just an unsupported configuration.

**Style note.** `ruff` is not available in the environment this was developed in,
so the 88-character limit was checked manually; the count of over-long lines in
both touched files is unchanged (28 → 28), i.e. no new violations. Worth a
`pre-commit run` on CI.

**Follow-ups this PR deliberately leaves out** (all recorded in the plan doc):

- An optional device-free layout-level assertion that the chosen output STL's
  `stride_map` derives from the *output* buffer (1024), not the input (1536) —
  a cheap invariant that would have caught this class of bug directly.
- Assert `rescale_stl_for_dtype`'s same-shape precondition inside the helper. Its
  docstring says "same-shape … dtype conversion", but no caller enforced it, which
  is how this bug was possible.
- Use `coordinates_normalizable` at the other candidate-filtering sites, so
  non-stick coordinate validation stops being a codegen-time surprise generally.

#### Does this PR introduce a user-facing change?

Yes — a correctness fix. Models whose dtype conversion reads a slice of a wider
buffer (Gemma per-head `q_norm`/`k_norm` under vLLM's native FP32 RMSNorm) either
failed to compile or returned wrong values; they now compile and match eager. No
API change and no change to any layout decision for a conversion that reads its
whole input.

#### Additional note:

cc @moriohara 

AI assistance was used for the investigation and for preparing this change.

### CODE REVIEW

**Comment by marnold-ibm** [comment]
This looks good other than the new test file not being wired up properly.  
Happy to approve once that's fixed.


### Summary

Fixes a layout-propagation bug in `_single_arg_op_layout` where a staggered dtype conversion inherits the input buffer's `device_size`/`stride_map` even when the conversion reads only a *slice* of that buffer. The fix gates the inherit path on `_convert_reads_whole_input()`, a new predicate that verifies the conversion accesses its input exactly as it writes its output. Sliced reads fall through to the existing dense reconstruction, which is physically sound.

The PR description is excellent — clear root-cause analysis, worked examples with concrete shapes, and honest disclosure of the verification gap on the large Gemma 4 checkpoints.

---

### Review

#### BLOCKERs

- **CI check failure: new test not wired into CI matrix** (`tests/inductor/test_sliced_staggered_convert.py`). The `All test_*.py files wired into GHA CI` check is already
failing. Two things are needed:
  1. A config YAML under `tests/configs/torch_spyre_tests/inductor/`, following the pattern of `test_ea_bidirectional_conversion_config.yaml`:
     ```yaml
     test_suite_config:
       labels: [unit, regression, integration, trunk]
       files:
         - path: ${TORCH_DEVICE_ROOT}/tests/inductor/test_sliced_staggered_convert.py
           unlisted_test_mode: mandatory_success
     ```
  2. A corresponding entry in `.github/workflows/_test_matrix.yaml`.

#### SUGGESTIONs

- **Test tolerance is loose** (`rtol=0.05, atol=0.05` in `test_sliced_staggered_convert.py`). Comparable tests in the repo use `atol=1e-2, rtol=1e-2`. The PR's observed maxabsdiff is 0.0078–0.0195, so tightening to e.g. `rtol=0.01, atol=0.03` would make the silent-wrong-results guard more sensitive without risk of spurious failures.

#### QUESTIONs

- The restickify-expansion block (line ~528 in `propagate_layouts.py`) that exposes alternative stick axes for STANDARD→staggered conversions now only runs on the `_convert_reads_whole_input` path. For a sliced STANDARD→DL16_TO_FP32 conversion that falls through to the dense path, that expansion is skipped. Is there a case where a sliced input still needs it, or is the answer that a sliced input's stick-axis choice can never be soundly inherited anyway?

- Is there a tracking issue for the follow-up `coordinates_normalizable` work at other candidate-filtering sites? Worth having so it doesn't get lost.

---

### Verdict

**REQUEST_CHANGES** — CI wiring is a hard blocker. The fix logic itself is correct and well-explained.

**Review by marnold-ibm — DISMISSED** [review]
Look good other than the new test file not being configured properly and failing.

REQUEST_CHANGES — CI wiring is a hard blocker. The fix logic itself is correct and well-explained.

**Comment by bohnstingl** [comment]
@marnold-ibm I should have now wired the test int correctly. On top of that, I also had to 

Fix the coarse-tiling symbol remapping for reductions whose loop variable does not appear in a memory index, such as a reduction over a broadcast value. The remapping now uses Inductor’s traced `ReadWrites.range_vars`, preserving the reduction symbol until it is intentionally removed when the tiled extent becomes 1.
Added a regression test for this symbol-remapping case.
This should fix the TP>1 Gemma4 tests in `spyre-inference`, see https://github.com/torch-spyre/spyre-inference/pull/845. Please have another look if you can.

**Comment by dgrove-oss** [comment]
Sorry to come to this late.  I think we should avoid a new test file/config entirely.

------

Fixes a layout-propagation bug where a staggered dtype conversion (DL16_TO_FP32/FP32_TO_DL16) that reads a slice of a wider buffer inherits the wrong row span from the input, causing either a hard codegen error or — worse — silently wrong results. Reached by vLLM's native FP32 RMSNorm on Gemma's per-head q_norm/k_norm. Fix gates the inherit-layout path on the conversion traversing its whole input (_convert_reads_whole_input) and adds a last-resort coordinates_normalizable filter.

Review

BLOCKERs

None.

SUGGESTIONs

- New test file is unnecessary; the coverage belongs in tests/inductor/test_ea_bidirectional_conversion.py. That file already targets exactly the code path this PR patches (the STANDARD↔DL16_TO_FP32/FP32_TO_DL16 EA-inheritance branch of _single_arg_op_layout), already has the right scaffolding (_run, assert_val, assert_ea/get_spyre_tensor_layout EA checks, dtype parametrization via DtypeOpTable.fp16_types()), and already contains a near-identical test (test_stagger_to_standard_ea) doing torch.compile + assert_close + an EA assertion for the non-sliced case. The new q_norm/k_norm cases in test_sliced_staggered_convert.py are a "sliced" variant of that same test and would fit as one or two more parametrized cases there with minimal new code — and would gain the EA-correctness assertion the new file's tests don't do at all (they only check values, not that the resulting layout is actually the staggered EA the fix is about). Notably, the PR's own commit for the coarse_tile.py change did add its regression case directly into the existing test_coarse_tiling.py rather than a new file — the same approach should apply here for consistency.
- The _convert_reads_whole_input docstring is thorough, but the function itself doesn't explain why dep.index == output_dep.index combined with equal host sizes is sufficient to prove "same access pattern" rather than, e.g., checking host_coordinates equality alone. Minor — the PR body's math derivation covers this, but a one-line comment tying the three conditions to the specific failure mode would help a future reader who hits this function without the PR description in hand.
- coordinates_normalizable is described as a "secondary, last-resort filter" but its use-site isn't shown/reviewed in isolation from _convert_reads_whole_input — the PR body already flags this as suggested future work to reuse elsewhere, which is reasonable to defer.

QUESTIONs

- The PR body explicitly flags a verification gap: the 4/5 geometry (Gemma-4 31B/26B-A4B) is validated standalone and at unit level but not end-to-end on real checkpoints. Is there a tracking issue for running that end-to-end confirmation, or is it expected to happen as a natural follow-up once #4238-adjacent Gemma work lands?

Verdict

APPROVE WITH SUGGESTIONS — the fix itself is well-reasoned and precisely scoped (verified against the PR's own math for the a==1/a!=1 cases), CI is green, DCO is valid, and the author is unusually transparent about scope and gaps. The one thing worth pushing back on before merge is exactly what you flagged: test_sliced_staggered_convert.py duplicates infrastructure that test_ea_bidirectional_conversion.py already has for this precise code path, and folding the new cases in there would give tighter, more idiomatic coverage (including an EA check the new tests currently lack) for less code.

**Comment by bohnstingl** [comment]
@marnold-ibm @dgrove-oss I should have addressed your feedback.

**Comment by bohnstingl** [comment]
I don't know whether the CI fails are related though

**Review by dgrove-oss — APPROVED** [review]
Approved.  Thanks.
